--- gnome-screensaver-2.28.3.orig/src/gnome-screensaver-preferences.c	2010-02-08 21:24:30.000000000 +0530
+++ gnome-screensaver-2.28.3.orig/src/gnome-screensaver-preferences.c	2011-11-29 18:21:50.000000000 +0530
@@ -1052,7 +1052,6 @@ 
         GtkWidget *widget;
 
         widget = GTK_WIDGET (gtk_builder_get_object (builder, "lock_checkbox"));
-        gtk_widget_set_sensitive (widget, !disable);
         if (disable) {
                 gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON (widget), FALSE);
         }

