diff -up gnome-screensaver-2.28.3/src/gs-grab-x11.c.movenotnuke gnome-screensaver-2.28.3/src/gs-grab-x11.c
--- gnome-screensaver-2.28.3/src/gs-grab-x11.c.movenotnuke	2013-02-05 13:36:28.571059681 -0500
+++ gnome-screensaver-2.28.3/src/gs-grab-x11.c	2013-02-05 13:36:35.034993875 -0500
@@ -395,7 +395,7 @@ gs_grab_move_keyboard (GSGrab    *grab,
 }
 
 static void
-gs_grab_nuke_focus (void)
+gs_grab_move_focus (GdkWindow *window)
 {
         Window focus = 0;
         int    rev = 0;
@@ -406,7 +406,7 @@ gs_grab_nuke_focus (void)
 
         XGetInputFocus (GDK_DISPLAY (), &focus, &rev);
 
-        XSetInputFocus (GDK_DISPLAY (), None, RevertToNone, CurrentTime);
+        XSetInputFocus (GDK_DISPLAY (), GDK_WINDOW_XID (window), RevertToNone, CurrentTime);
 
         gdk_display_sync (gdk_display_get_default ());
         gdk_error_trap_pop ();
@@ -454,7 +454,7 @@ gs_grab_grab_window (GSGrab    *grab,
         if (kstatus != GDK_GRAB_SUCCESS) {
                 if (!focus_fuckus) {
                         focus_fuckus = TRUE;
-                        gs_grab_nuke_focus ();
+                        gs_grab_move_focus (window);
                         goto AGAIN;
                 }
         }
